README.md
#EgretI18n  
皮肤文件中绑定数据的多语言支持

绑定数据格式需为： $i18n.\*

#命令：  
egreti18n save ： 保存exml.json文件中绑定数据$i18n.\* 至 resource/i18n/{language} 文件夹  
egreti18n use  ： 合并resource/i18n/{language}中的数据至 exml.json($i18n.\*) 并切换语言  
extension.EgretI18nExportChinese： 导出皮肤文件中的中文到exml.json文件  
#设置：  
EgretI18n.skinPath：  设置皮肤所在文件夹，默认值： resoruce/skins  
EgretI18n.exportTags： 皮肤文件中要导出中文的标签  
```
默认值 
[
   ["e:Label", "text"],
   ["e:Label", "text.up"],
   ["e:Label", "text.down"],
   ["e:Label", "text.disabled"],
   ["e:Label", "text.unOpen"],
   ["e:BitmapLabel", "text"],
   ["e:Button", "label"],
   ["e:EditableText", "prompt"],
   ["e:EditableText", "text"],
   ["e:TextInput", "prompt"]
 ]
 ```

CHANGELOG.md
# Change Log

All notable changes to the "egreti18n" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]

- Initial release

src/save.ts
import {
  findFile,
  formattime,
  readFile,
  shallowEqual,
  writeFile,
  Log
} from "./util";

const fs = require("fs");
const vscode = require("vscode");
const childProcess = require("child_process");

let name_id: { [x: string]: string } = {};
let id_name: { [x: string]: string } = {};
let existPaths: { [x: string]: boolean } = {};

const workPath = vscode.workspace.rootPath;
const exmlPath = ".wing/exml.json";
let skinPath = "";
const destPath = "resource/i18n";
const cachePath = "temp/i18n";

Log.appendLine("工作路径：" + workPath);

function operateExmlfile(p: string) {
  return readFile(p, "utf-8")
    .then((data: any) => {
      let idstr = data.match(/w:Config id="[0-9a-zA-Z]+"/);
      if (!idstr || idstr.length <= 0) {
        return;
      }
      let id: any = /"[0-9a-zA-Z]+"/.exec(idstr[0]);
      id = id && id[0].replace(/"/g, "");
      // let temp = p.split("/");
      // temp = temp[temp.length - 1].split(".");
      // let name = temp[0];

      if (id_name[id]) {
        Log.appendLine("修改重复id  " + p.replace(workPath, ""));
        while (id_name[id]) {
          id = (Date.now() - Math.round(Math.random() * 10000000)).toString(16);
        }
        writeFile(
          p,
          data.replace(/w:Config id="[0-9a-zA-Z]+"/, `w:Config id="${id}"`)
        );
      }
      id_name[id] = p;
      name_id[p] = id;
    })
    .catch((e: any) => {
      Log.appendLine("失败operateExmlfile：" + e.message);
    });
}

//读取皮肤文件结束
function readSkinFiles() {
  return findFile(`${workPath}/${skinPath}`, /\.exml$/).then(files => {
    if (!files || files.length <= 0) {
      Log.appendLine(`${skinPath}中找不到.exml文件，请设置`);
      throw new Error(`${skinPath}中找不到.exml文件，请设置`);
    }
    let promises: Promise<any>[] = [];
    files.forEach(file => {
      promises.push(operateExmlfile(file));
    });
    return Promise.all(promises).then(() => {
      Log.appendLine("读取皮肤文件结束！");
    });
  });
}

function checkFilePath(p: string) {
  if (existPaths[p]) {
    return;
  }
  if (fs.existsSync(p)) {
    existPaths[p] = true;
    return;
  }
  let paths = p.split("/");
  let ps: string[] = [];
  paths.forEach(item => {
    ps.push(item);
    let temppath = ps.join("/");

    if (existPaths[temppath] || workPath.indexOf(temppath) != -1) {
      return;
    }
    if (fs.existsSync(temppath)) {
      existPaths[temppath] = true;
      return;
    }
    Log.appendLine("mkdir：" + temppath);
    fs.mkdirSync(temppath);
    existPaths[temppath] = true;
  });
}

function parseExmlFile(lan: string): Promise<{ [x: string]: any }> {
  let path = `${workPath}/${exmlPath}`;
  return readFile(path, "utf-8")
    .then((data: string) => {
      let obj = JSON.parse(data);
      Log.appendLine(`解析${path}成功`);
      return obj;
    })
    .catch((e: any) => {
      throw new Error(`解析${path}失败` + e.message);
    });
}

function writeJsonFile(data: { [x: string]: any }, lan: string) {
  // let promises = [];
  let output: any = {};
  for (let key in data) {
    const item = data[key];
    if (typeof item === "object") {
      let bindingDataTestObj = item.bindingDataTestObj;
      //exml文件路径
      let filePath = id_name[key];
      if (bindingDataTestObj && bindingDataTestObj.length > 0 && filePath) {
        //目标路径
        // let dPath = filePath
        //   .replace(skinPath, destPath + "/" + lan)
        //   .replace(/\/[0-9a-zA-Z_]+\.exml$/, "");

        // let arrF = filePath.split("/");
        let fileName = filePath
          .replace(workPath + "/" + skinPath + "/", "")
          .replace(".exml", "")
          .replace(/\//g, "_");

        //文本内容
        let content: { [x: string]: string } = {};
        bindingDataTestObj.forEach((item: any) => {
          item.key.indexOf("$i18n.") != -1 &&
            (content[item.key.replace("$i18n.", "")] = item.value);
        });
        if (Object.keys(content).length > 0) {
          output[fileName] = content;
          //   //创建文件夹
          //   checkFilePath(dPath);

          //   let filePath = dPath + "/" + fileName + "_" + lan + ".json";
          //   console.log(filePath);
          //   //有数据---存储文件
          //   let _promise = writeFile(filePath, JSON.stringify(content)).then(
          //     () => {
          //       Log.appendLine(
          //         "更新文件：" + filePath.replace(workPath + "/", "")
          //       );
          //     }
          //   );
          //   promises.push(_promise);
        }
      }
    }
  }
  checkFilePath(`${workPath}/${destPath}`);
  return writeFile(
    `${workPath}/${destPath}/${lan}.json`,
    JSON.stringify(output)
  );

  // return Promise.all(promises);
}

/**
 * 开始
 */
export function save(lan: string) {
  Log.show();

  skinPath =
    vscode.workspace.getConfiguration().get("EgretI18n.skinPath") ||
    "resource/skins";
  Log.appendLine("EgretI18n.skinPath 皮肤文件路径：" + skinPath);

  let cache = lan + formattime(Date.now());
  // checkFilePath(`${workPath}/${destPath}`);
  // checkFilePath(`${workPath}/${cachePath}`);
  Log.appendLine("备份文件至：" + cache);
  const p = childProcess.exec(`
  mv ${workPath}/${destPath}/${lan}.json ${workPath}/${cachePath}/${cache}.json
  `);
  p.once("exit", function(code: number) {
    // if (code != 0) {
    // Log.appendLine("备份文件失败：" + code);
    // vscode.window.showInformationMessage("备份文件失败" + code);
    // return;
    // }
    existPaths = {};

    Log.appendLine("开始读取皮肤文件！");

    readSkinFiles()
      .then(() => parseExmlFile(lan))
      .then(data => writeJsonFile(data, lan))
      .then(() => {
        Log.appendLine("success!");
        vscode.window.showInformationMessage("成功");
      })
      .catch((e: any) => {
        Log.appendLine("Fail  " + e.message);
        vscode.window.showErrorMessage("失败");
      })
      .then(() => {
        //清空
        name_id = {};
        id_name = {};
        existPaths = {};
      });
  });
}


src/util.ts
const fs = require("fs");
const Path = require("path");
const vscode = require("vscode");

export function promisify(fn: Function): Function {
  return function(...args: any): Promise<any> {
    return new Promise((resolve, reject) => {
      fn(...args, (err: any, data: any) => {
        if (err) {
          reject(err);
        } else {
          resolve(data);
        }
      });
    });
  };
}

export const stat = promisify(fs.stat);
export const readdir = promisify(fs.readdir);
export const readFile = promisify(fs.readFile);
export const writeFile = promisify(fs.writeFile);

export function isDir(filePath: string) {
  return stat(filePath)
    .then((stats: any) => stats.isDirectory())
    .catch(() => false);
}

export function findFile(filePath: string, regexp: RegExp): Promise<string[]> {
  return isDir(filePath).then((isDirectory: boolean) => {
    if (!isDirectory) {
      return [];
    }
    return readdir(filePath).then((files: string[]) => {
      let promises = [];
      let paths: string[] = [];
      for (let file of files) {
        const fullPath = Path.join(filePath, file);

        let _promise = isDir(fullPath).then((isDirectory: boolean) => {
          if (isDirectory) {
            return findFile(fullPath, regexp).then(result => {
              paths = paths.concat(result);
            });
          } else if (regexp.test(fullPath)) {
            paths.push(fullPath);
          }
        });
        promises.push(_promise);
      }
      return Promise.all(promises).then(() => paths);
    });
  });
}

function is(x: any, y: any) {
  // From: https://github.com/facebook/fbjs/blob/c69904a511b900266935168223063dd8772dfc40/packages/fbjs/src/core/shallowEqual.js
  if (x === y) {
    return x !== 0 || 1 / x === 1 / y;
  } else {
    return x !== x && y !== y;
  }
}

export function shallowEqual(objA: any, objB: any) {
  //From: https://github.com/facebook/fbjs/blob/c69904a511b900266935168223063dd8772dfc40/packages/fbjs/src/core/shallowEqual.js
  if (is(objA, objB)) return true;
  if (
    typeof objA !== "object" ||
    objA === null ||
    typeof objB !== "object" ||
    objB === null
  ) {
    return false;
  }
  const keysA = Object.keys(objA);
  const keysB = Object.keys(objB);
  if (keysA.length !== keysB.length) return false;
  for (let i = 0; i < keysA.length; i++) {
    if (
      !Object.prototype.hasOwnProperty.call(objB, keysA[i]) ||
      !is(objA[keysA[i]], objB[keysA[i]])
    ) {
      return false;
    }
  }
  return true;
}

export function coverData(arrA: Array<any>, arrB: Array<any>) {
  let da: { [x: string]: string } = {};
  let db: { [x: string]: string } = {};
  arrA && arrA.forEach(item => (da[item.key] = item.value));
  arrB && arrB.forEach(item => (db[item.key] = item.value));
  da = { ...da, ...db };
  let arr = [];
  for (let key in da) {
    arr.push({
      key,
      value: da[key]
    });
  }
  return arr;
}

export function formattime(time: number, formatstr = "yyyy-MM-dd-HH:mm:ss") {
  let date = new Date(time);
  return formatstr
    .replace(/yyyy/gi, date.getFullYear().toString())
    .replace(/MM/g, (date.getMonth() + 1).toString())
    .replace(/DD/gi, date.getDate().toString())
    .replace(/hh/gi, date.getHours().toString())
    .replace(/mm/g, date.getMinutes().toString())
    .replace(/ss/gi, date.getSeconds().toString());
}

export const Log = vscode.window.createOutputChannel("egreti18n");


src/useLan.ts
import {
  findFile,
  readdir,
  readFile,
  writeFile,
  shallowEqual,
  coverData,
  Log
} from "./util";

const fs = require("fs");
const vscode = require("vscode");

let name_id: { [x: string]: string } = {};
let id_name: { [x: string]: string } = {};
let existPaths: { [x: string]: boolean } = {};

const workPath = vscode.workspace.rootPath;
let skinPath = "";
const destPath = ".wing";
const jsonPath = "resource/i18n";

let output: any = {};

Log.appendLine("工作路径：" + workPath);

function operateExmlfile(p: string) {
  return readFile(p, "utf-8")
    .then((data: any) => {
      let idstr = data.match(/w:Config id="[0-9a-zA-Z]+"/);
      if (!idstr || idstr.length <= 0) {
        return;
      }
      let id: any = /"[0-9a-zA-Z]+"/.exec(idstr[0]);
      id = id && id[0].replace(/"/g, "");
      // let temp = p.split("/");
      // temp = temp[temp.length - 1].split(".");
      // let name = temp[0];

      if (id_name[id]) {
        Log.appendLine("修改重复id  " + p.replace(workPath, ""));

        while (id_name[id]) {
          id = (Date.now() - Math.round(Math.random() * 10000000)).toString(16);
        }
        writeFile(
          p,
          data.replace(/w:Config id="[0-9a-zA-Z]+"/, `w:Config id="${id}"`)
        );
      }
      id_name[id] = p;
      name_id[
        p.replace(workPath + "/" + skinPath + "/", "").replace(/\//g, "_")
      ] = id;
    })
    .catch((e: any) => {
      Log.appendLine("失败operateExmlfile：" + e.message);
    });
}
//读取皮肤文件结束
function readSkinFiles() {
  return findFile(`${workPath}/${skinPath}`, /\.exml$/).then(files => {
    // Log.appendLine("读取皮肤文件结束！");
    let promises: Promise<any>[] = [];
    files.forEach(file => {
      promises.push(operateExmlfile(file));
    });
    return Promise.all(promises).then(() => {
      Log.appendLine("读取皮肤文件结束！");
    });
  });
}

function readLanFiles(lan: string): Promise<any> {
  output[lan] = {};
  // return findFile(`${workPath}/${jsonPath}/${lan}`, /\.json$/).then(
  //   (paths: string[]) => {
  //     if (!paths || paths.length <= 0) {
  //       throw new Error(`找不到 ${workPath}/${jsonPath}/${lan}`);
  //     }
  //     let promises: Promise<any>[] = [];
  //     paths.forEach((path: string) => {
  return readFile(`${workPath}/${jsonPath}/${lan}.json`, "utf-8").then(
    (str: string) => {
      let data = JSON.parse(str);
      for (let skinname in data) {
        let bindingDataTestObj = [];
        for (let key in data[skinname]) {
          bindingDataTestObj.push({
            key: `$i18n.${key}`,
            value: data[skinname][key]
          });
        }
        if (bindingDataTestObj.length > 0 && name_id[skinname + ".exml"]) {
          output[lan][name_id[skinname + ".exml"]] = {
            bindingDataTestObj
          };
        }
      }
    }
  );

  //       promises.push(_promise);
  //     });
  //     return Promise.all(promises);
  //   }
  // );
}

function writeJsonFile(lan: string) {
  let jsonPath = `${workPath}/${destPath}/exml.json`;
  return readFile(jsonPath, "utf-8")
    .catch(() => JSON.stringify({}))
    .then((fileStr: string) => {
      let data: any = {};
      try {
        data = JSON.parse(fileStr);
      } catch (e) {}
      for (let key in output[lan]) {
        data[key] = data[key] || {};
        data[key]["bindingDataTestObj"] = coverData(
          data[key]["bindingDataTestObj"],
          output[lan][key]["bindingDataTestObj"]
        );
      }
      return writeFile(jsonPath, JSON.stringify(data)).then(() => {
        Log.appendLine(`更新文件：${jsonPath}`);
      });
    })
    .catch((e: any) => {
      Log.appendLine(`更新文件失败：${jsonPath}, ${e.message}`);
    });
}

function checkFilePath(p: string) {
  if (existPaths[p]) {
    return;
  }
  if (fs.existsSync(p)) {
    existPaths[p] = true;
    return;
  }
  let paths = p.split("/");
  let ps: string[] = [];
  paths.forEach(item => {
    ps.push(item);
    let temppath = ps.join("/");

    if (existPaths[temppath] || workPath.indexOf(temppath) != -1) {
      return;
    }
    if (fs.existsSync(temppath)) {
      existPaths[temppath] = true;
      return;
    }
    Log.appendLine("mkdir：" + temppath);
    fs.mkdirSync(temppath);
    existPaths[temppath] = true;
  });
}

/**
 * 开始
 */
export function useLan(lan: string) {
  Log.show();

  skinPath =
    vscode.workspace.getConfiguration().get("EgretI18n.skinPath") ||
    "resource/skins";
  Log.appendLine("EgretI18n.skinPath 皮肤文件路径：" + skinPath);

  Log.appendLine("开始读取皮肤文件！");
  return readSkinFiles()
    .then(() => readLanFiles(lan))
    .then(() => writeJsonFile(lan))
    .then(() => {
      Log.appendLine("success");
      //清空
      name_id = {};
      id_name = {};
      existPaths = {};
      output = {};
    })
    .catch((e: any) => {
      Log.appendLine("Fail " + e.message);
      name_id = {};
      id_name = {};
      existPaths = {};
      output = {};
      vscode.window.showErrorMessage("失败");
      throw "err";
    });
}


src/extension.ts
// The module 'vscode' contains the VS Code extensibility API
// Import the module and reference it with the alias vscode in your code below
import * as vscode from "vscode";
// import * as Exml2Json from "./exml2json";
// import * as Json2Exml from "./json2exml";

import { save } from "./save";
import { useLan } from "./useLan";
import { exportChinese } from "./exportChinese";

export function activate(context: vscode.ExtensionContext) {
  console.log("egreti18n start");
  // let disposable = vscode.commands.registerCommand(
  //   "extension.egreti18nexml2json",
  //   () => {
  //     vscode.window.showInformationMessage("egret i18n exml2json");
  //     Exml2Json.start();
  //   }
  // );
  // context.subscriptions.push(disposable);
  // disposable = vscode.commands.registerCommand(
  //   "extension.egreti18njson2exml",
  //   () => {
  //     vscode.window.showInformationMessage("egret i18n json2exml");
  //     Json2Exml.start();
  //   }
  // );
  // context.subscriptions.push(disposable);
  //保存exml数据至json
  let disposable = vscode.commands.registerCommand(
    "extension.EgretI18nSave",
    () => {
      vscode.window.showInputBox().then(lan => {
        if (!lan) {
          vscode.window.showInformationMessage("请输入要保存至的语言类型");
          return;
        }
        vscode.window.showInformationMessage("保存文件至resource/i18n，覆盖");
        save(lan);
      });
    }
  );
  context.subscriptions.push(disposable);
  //使用语言
  disposable = vscode.commands.registerCommand("extension.EgretI18nUse", () => {
    vscode.window.showInputBox().then(lan => {
      if (!lan) {
        vscode.window.showInformationMessage("请输入要切换的语言类型");
        return;
      }
      vscode.window.showInformationMessage("切换语言，合并至.wing/exmljson");
      useLan(lan).then(() => {
        vscode.window.showInformationMessage("切换语言成功，重载wing后生效");

        // setTimeout(
        //   () => vscode.commands.executeCommand("workbench.action.reloadWindow"),
        //   1000
        // );
      });
    });
  });
  //导出皮肤中的中文标签
  disposable = vscode.commands.registerCommand(
    "extension.EgretI18nExportChinese",
    () => {
      exportChinese();
    }
  );
  context.subscriptions.push(disposable);
}

// this method is called when your extension is deactivated
export function deactivate() {}


src/json2exml.ts
import {
  findFile,
  readdir,
  readFile,
  writeFile,
  shallowEqual,
  coverData
} from "./util";

const fs = require("fs");
const vscode = require("vscode");

let name_id: { [x: string]: string } = {};
let id_name: { [x: string]: string } = {};
let existPaths: { [x: string]: boolean } = {};

const workPath = vscode.workspace.rootPath;
const skinPath = "resource/skins";
const destPath = ".wing";
const jsonPath = "resource/i18n";

let output: any = {};

let Log = vscode.window.createOutputChannel("egreti18n");

Log.appendLine("工作路径：" + workPath);

function operateExmlfile(p: string) {
  return readFile(p, "utf-8")
    .then((data: any) => {
      let idstr = data.match(/w:Config id="[0-9a-zA-Z]+"/);
      if (!idstr || idstr.length <= 0) {
        return;
      }
      let id: any = /"[0-9a-zA-Z]+"/.exec(idstr[0]);
      id = id && id[0].replace(/"/g, "");
      // let temp = p.split("/");
      // temp = temp[temp.length - 1].split(".");
      // let name = temp[0];

      if (id_name[id]) {
        Log.appendLine("修改重复id  " + p.replace(workPath, ""));

        while (id_name[id]) {
          id = (Date.now() - Math.round(Math.random() * 10000000)).toString(16);
        }
        writeFile(
          p,
          data.replace(/w:Config id="[0-9a-zA-Z]+"/, `w:Config id="${id}"`)
        );
      }
      id_name[id] = p;
      name_id[p] = id;
    })
    .catch((e: any) => {
      Log.appendLine("失败operateExmlfile：" + e.message);
    });
}
//读取皮肤文件结束
function readSkinFiles() {
  return findFile(`${workPath}/${skinPath}`, /\.exml$/).then(files => {
    // Log.appendLine("读取皮肤文件结束！");
    let promises: Promise<any>[] = [];
    files.forEach(file => {
      promises.push(operateExmlfile(file));
    });
    return Promise.all(promises).then(() => {
      Log.appendLine("读取皮肤文件结束！");
    });
  });
}

function readJsonFiles() {
  return readdir(`${workPath}/${jsonPath}`).then((files: string[]) => {
    let promises: Promise<any>[] = [];
    files.forEach((lan: string) => {
      promises.push(readLanFiles(lan).then(() => writeJsonFile(lan)));
    });
    return Promise.all(promises);
  });
}

function readLanFiles(lan: string): Promise<any> {
  output[lan] = {};
  return findFile(`${workPath}/${jsonPath}/${lan}`, /\.json$/).then(
    (paths: string[]) => {
      let promises: Promise<any>[] = [];
      paths.forEach((path: string) => {
        let _promise = readFile(path, "utf-8").then((str: string) => {
          let data = JSON.parse(str);
          let skinname = path
            .replace(`${jsonPath}/${lan}`, skinPath)
            .replace("_" + lan + ".json", ".exml");
          let bindingDataTestObj = [];
          for (let key in data) {
            bindingDataTestObj.push({
              key: `$i18n.${key}`,
              value: data[key]
            });
          }
          if (bindingDataTestObj.length > 0 && name_id[skinname]) {
            output[lan][name_id[skinname]] = {
              bindingDataTestObj
            };
          }
        });
        promises.push(_promise);
      });
      return Promise.all(promises);
    }
  );
}

function writeJsonFile(lan: string) {
  let jsonPath = `${workPath}/${destPath}/exml.${lan}.json`;
  return readFile(jsonPath, "utf-8")
    .catch(() => JSON.stringify({}))
    .then((fileStr: string) => {
      let data = JSON.parse(fileStr);
      for (let key in output[lan]) {
        data[key] = data[key] || {};
        data[key]["bindingDataTestObj"] = coverData(
          data[key]["bindingDataTestObj"],
          output[lan][key]["bindingDataTestObj"]
        );
      }
      return writeFile(jsonPath, JSON.stringify(data)).then(() => {
        Log.appendLine(`更新文件：${workPath}/${destPath}/exml.${lan}.json`);
      });
    })
    .catch(() => {
      Log.appendLine(`更新文件失败：${workPath}/${destPath}/exml.${lan}.json`);
    });
}

function checkFilePath(p: string) {
  if (existPaths[p]) {
    return;
  }
  if (fs.existsSync(p)) {
    existPaths[p] = true;
    return;
  }
  let paths = p.split("/");
  let ps: string[] = [];
  paths.forEach(item => {
    ps.push(item);
    let temppath = ps.join("/");

    if (existPaths[temppath] || workPath.indexOf(temppath) != -1) {
      return;
    }
    if (fs.existsSync(temppath)) {
      existPaths[temppath] = true;
      return;
    }
    Log.appendLine("mkdir：" + temppath);
    fs.mkdirSync(temppath);
    existPaths[temppath] = true;
  });
}

/**
 * 开始
 */
export function start() {
  Log.show();
  Log.appendLine("开始读取皮肤文件！");

  readSkinFiles()
    .then(readJsonFiles)
    .then(() => {
      Log.appendLine("success");
    })
    .catch((e: any) => {
      Log.appendLine("Fail", e.message);
    })
    .then(() => {
      //清空
      name_id = {};
      id_name = {};
      existPaths = {};
      output = {};
    });
  // });
}


src/exportChinese.ts
import { findFile, readFile, writeFile, coverData, Log } from "./util";

const vscode = require("vscode");
const uuid = require("uuid");

const workPath = vscode.workspace.rootPath;

let skinPath = "";
let exmlPath = ".wing/exml.json";

let output: { [x: string]: any } = {};
let id_name: { [x: string]: string } = {};

//读取皮肤文件结束
function readSkinFiles() {
  return findFile(`${workPath}/${skinPath}`, /\.exml$/).then(
    (files: string[]) => {
      if (!files || files.length <= 0) {
        throw new Error(`${skinPath}中找不到.exml文件，请设置`);
      }
      let promises: Promise<any>[] = [];
      files.forEach(file => {
        promises.push(operateExmlfile(file));
      });
      return Promise.all(promises).then(() => {
        console.log("读取皮肤文件结束！");
      });
    }
  );
}

function operateExmlfile(p: string): Promise<any> {
  return readFile(p, "utf-8")
    .then((originalData: any) => {
      let data = originalData;
      let idstr = data.match(/w:Config id="[0-9a-zA-Z]+"/);
      let id: any = "";
      if (!idstr || idstr.length <= 0) {
        id = (Date.now() - Math.round(Math.random() * 10000000)).toString(16);
        data = data.replace(/<\?xml .+\?>[\n]+<e:Skin [^>]+>/, (v: string) => {
          if (!/xmlns:w=".+"/.test(v)) {
            v =
              v.substr(0, v.length - 1) +
              ` xmlns:w="http://ns.egret.com/wing">`;
          }
          return `${v}
  <w:Config id="${id}"/>`;
        });
      } else {
        id = /"[0-9a-zA-Z]+"/.exec(idstr[0]);
        id = id && id[0].replace(/"/g, "");
      }

      if (id_name[id]) {
        while (id_name[id]) {
          id = (Date.now() - Math.round(Math.random() * 10000000)).toString(16);
        }
        data = data.replace(
          /w:Config id="[0-9a-zA-Z]+"/,
          `w:Config id="${id}"`
        );
        id_name[id] = p;
      }

      vscode.workspace
        .getConfiguration()
        .get("EgretI18n.exportTags")
        .forEach(([tag, lab]: [string, string]) => {
          data = data.replace(
            eval(`/<${tag} ((?!>[ ]*\\n)[\\s\\S])*>/g`),
            (v: string) => {
              return v.replace(eval(`/${lab}="[^"]+"/`), (content: string) => {
                if (!/[\u4e00-\u9fa5]/.test(content)) {
                  return content;
                }
                if (eval(`/${lab}="{.*}"/`).test(content)) {
                  return content;
                } else {
                  let slashLen = content.match(/"/g);
                  slashLen && slashLen.length > 2 && console.warn(content, p);
                  /[{}]/.test(content) && console.warn(content, p);
                  eval(`/${lab}=""/`).test(content) && console.warn(content, p);

                  let value = content
                    .replace(eval(`/^${lab}="/`), "")
                    .replace(/"$/, "");
                  output[id] = output[id] || [];
                  let uid = "";
                  // while (true) {
                  uid = uuid.v4().replace(/-/g, "");
                  //   let isUnique = output[id].every(item => item.key !== `$i18n.${uid}`);
                  //   if (isUnique) {
                  //     break;
                  //   }
                  // }
                  for (let i = 0; i < output[id].length; i++) {
                    if (output[id][i].value === value) {
                      return `${lab}="{${output[id].key}}"`;
                    }
                  }
                  output[id].push({
                    key: `$i18n.${uid}`,
                    value
                  });
                  return `${lab}="{$i18n.${uid}}"`;
                }
              });
            }
          );
        });

      return [data, originalData];
    })
    .then(([changeData, originalData]: [string, string]) => {
      if (changeData !== originalData) {
        return writeFile(p, changeData);
      }
    });
}

export function exportChinese() {
  Log.show();

  skinPath =
    vscode.workspace.getConfiguration().get("EgretI18n.skinPath") ||
    "resource/skins";
  Log.appendLine("EgretI18n.skinPath 皮肤文件路径：" + skinPath);

  Log.appendLine(
    "导出标签：" +
      JSON.stringify(
        vscode.workspace.getConfiguration().get("EgretI18n.exportTags")
      )
  );

  Log.appendLine("开始读取皮肤文件！");
  readSkinFiles()
    .then(() =>
      readFile(`${workPath}/${exmlPath}`, "utf-8").then((data: string) =>
        JSON.parse(data)
      )
    )
    .then(data => {
      for (let key in output) {
        data[key] = data[key] || {};
        data[key]["bindingDataTestObj"] = coverData(
          data[key]["bindingDataTestObj"],
          output[key]
        );
      }
      return writeFile(`${workPath}/${exmlPath}`, JSON.stringify(data)).then(
        () => {
          Log.appendLine(`更新文件：${exmlPath}`);
        }
      );
    })
    .catch(e => console.error("error:", e));
}


src/exml2json.ts
import { findFile, readFile, shallowEqual, writeFile } from "./util";

const fs = require("fs");
const vscode = require("vscode");

let name_id: { [x: string]: string } = {};
let id_name: { [x: string]: string } = {};
let existPaths: { [x: string]: boolean } = {};

const workPath = vscode.workspace.rootPath;
const exmlPath = ".wing";
const skinPath = "resource/skins";
const destPath = "resource/i18n";

let Log = vscode.window.createOutputChannel("egreti18n");

Log.appendLine("工作路径：" + workPath);

function operateExmlfile(p: string) {
  return readFile(p, "utf-8")
    .then((data: any) => {
      let idstr = data.match(/w:Config id="[0-9a-zA-Z]+"/);
      if (!idstr || idstr.length <= 0) {
        return;
      }
      let id: any = /"[0-9a-zA-Z]+"/.exec(idstr[0]);
      id = id && id[0].replace(/"/g, "");
      // let temp = p.split("/");
      // temp = temp[temp.length - 1].split(".");
      // let name = temp[0];

      if (id_name[id]) {
        Log.appendLine("修改重复id  " + p.replace(workPath, ""));
        while (id_name[id]) {
          id = (Date.now() - Math.round(Math.random() * 10000000)).toString(16);
        }
        writeFile(
          p,
          data.replace(/w:Config id="[0-9a-zA-Z]+"/, `w:Config id="${id}"`)
        );
      }
      id_name[id] = p;
      name_id[p] = id;
    })
    .catch((e: any) => {
      Log.appendLine("失败operateExmlfile：" + e.message);
    });
}

//读取皮肤文件结束
function readSkinFiles() {
  return findFile(`${workPath}/${skinPath}`, /\.exml$/).then(files => {
    // Log.appendLine("读取皮肤文件结束！");
    let promises: Promise<any>[] = [];
    files.forEach(file => {
      promises.push(operateExmlfile(file));
    });
    return Promise.all(promises).then(() => {
      Log.appendLine("读取皮肤文件结束！");
    });
  });
}

function checkFilePath(p: string) {
  if (existPaths[p]) {
    return;
  }
  if (fs.existsSync(p)) {
    existPaths[p] = true;
    return;
  }
  let paths = p.split("/");
  let ps: string[] = [];
  paths.forEach(item => {
    ps.push(item);
    let temppath = ps.join("/");

    if (existPaths[temppath] || workPath.indexOf(temppath) != -1) {
      return;
    }
    if (fs.existsSync(temppath)) {
      existPaths[temppath] = true;
      return;
    }
    Log.appendLine("mkdir：" + temppath);
    fs.mkdirSync(temppath);
    existPaths[temppath] = true;
  });
}

function parseExmlFiles() {
  return findFile(`${workPath}/${exmlPath}`, /exml\.[a-zA-Z]+\.json/).then(
    (files: string[]) => {
      let promises: any[] = [];
      files.forEach(file => {
        let arr = file.match(/\.[a-zA-Z]+\.json$/);
        let lan: any =
          arr && arr[0] && arr[0].replace(".json", "").replace(".", "");
        if (!lan) {
          return;
        }
        promises.push(
          parseExmlFile(lan, file).then(data => writeJsonFile(data, lan))
        );
      });

      return Promise.all(promises);
    }
  );
}

function parseExmlFile(
  lan: string,
  path: string
): Promise<{ [x: string]: any }> {
  return readFile(path, "utf-8")
    .then((data: string) => {
      let obj = JSON.parse(data);
      Log.appendLine(`解析${path}成功`);
      return obj;
    })
    .catch((e: any) => {
      throw new Error(`解析${path}失败` + e.message);
    });
}

function writeJsonFile(data: { [x: string]: any }, lan: string) {
  let promises = [];
  for (let key in data) {
    const item = data[key];
    if (typeof item === "object") {
      let bindingDataTestObj = item.bindingDataTestObj;
      //exml文件路径
      let filePath = id_name[key];
      if (bindingDataTestObj && bindingDataTestObj.length > 0 && filePath) {
        //目标路径
        let dPath = filePath
          .replace(skinPath, destPath + "/" + lan)
          .replace(/\/[0-9a-zA-Z_]+\.exml$/, "");

        let arrF = filePath.split("/");
        let fileName = arrF[arrF.length - 1].replace(".exml", "");

        //创建文件夹
        checkFilePath(dPath);
        //文本内容
        let content: { [x: string]: string } = {};
        bindingDataTestObj.forEach((item: any) => {
          item.key.indexOf("$i18n.") != -1 &&
            (content[item.key.replace("$i18n.", "")] = item.value);
        });
        if (Object.keys(content).length > 0) {
          let filePath = dPath + "/" + fileName + "_" + lan + ".json";
          //有数据---存储文件
          let _promise = readFile(filePath, "utf-8")
            .catch(() => JSON.stringify({}))
            .then((filecontent: string) => {
              let data = JSON.parse(filecontent);
              if (shallowEqual(data, content)) {
                Log.appendLine(
                  "无需更新：" + filePath.replace(workPath + "/", "")
                );
                return;
              }
              content = { ...data, ...content };
              return writeFile(filePath, JSON.stringify(content)).then(() => {
                Log.appendLine(
                  "更新文件：" + filePath.replace(workPath + "/", "")
                );
              });
            })
            .catch((e: any) => {
              Log.appendLine(
                "更新文件失败：" + filePath.replace(workPath + "/", "")
              );
            });
          promises.push(_promise);
        }
      }
    }
  }

  return Promise.all(promises);
}

/**
 * 开始
 */
export function start() {
  Log.show();
  Log.appendLine("开始读取皮肤文件！");

  readSkinFiles()
    .then(parseExmlFiles)
    .then(() => {
      Log.appendLine("success!");
    })
    .catch((e: any) => {
      Log.appendLine("Fail" + e.message);
    })
    .then(() => {
      //清空
      name_id = {};
      id_name = {};
      existPaths = {};
    });
  // });
}
